File: internal/serialsupply/contract.go
// AUTHORED bounded contracts for connecting existing procurement, factory and
// inventory owners. No new pricing, stock ledger or corporate implementation.
package serialsupply

import (
	"encoding/json"
	"errors"
	"regexp"
	"strings"
	"time"
	"unicode"
	"unicode/utf8"

	"elite.local/enterprise/internal/approval"
)

var ErrInvalid = errors.New("invalid serial supply request")
var ErrNotFound = errors.New("serial supply not found")
var ErrConflict = errors.New("serial supply conflict")
var idPattern = regexp.MustCompile(`^[A-Za-z0-9][A-Za-z0-9._:-]{0,127}$`)
var shaPattern = regexp.MustCompile(`^[0-9a-f]{64}$`)

const PolicyCode = "strict-serial-reference/v1"
const PolicyJSON = `{"schema":"elite-serial-supply-policy/v1","scope":"LIBRARY_INFRASTRUCTURE_REFERENCE","quantity":"whole-serialized-units","overreceipt":0,"receipt_state":"quarantine","release":"distinct-human-with-evidence","pricing":"existing-purchase-order-total-unchanged","production_authorized":false}`

func ValidID(s string) bool  { return idPattern.MatchString(s) }
func ValidSHA(s string) bool { return shaPattern.MatchString(s) }
func ValidText(s string, max int) bool {
	if s == "" || len(s) > max || !utf8.ValidString(s) || strings.TrimSpace(s) != s {
		return false
	}
	for _, r := range s {
		if unicode.IsControl(r) {
			return false
		}
	}
	return true
}
func Canonical(value any) (json.RawMessage, string, error) {
	raw, err := json.Marshal(value)
	if err != nil {
		return nil, "", err
	}
	out, hash, err := approval.CanonicalPayload(raw)
	return out, hash, err
}

type Line struct {
	ID        string `json:"id"`
	VariantID string `json:"variant_id"`
	Quantity  int    `json:"quantity"`
}
type PlanRequest struct {
	PurchaseOrderID       string `json:"purchase_order_id"`
	CommandID             string `json:"command_id"`
	FactoryOrganizationID string `json:"factory_organization_id"`
	DemandReference       string `json:"demand_reference"`
	PolicyCode            string `json:"policy_code"`
	EvidenceSHA256        string `json:"evidence_sha256"`
	Lines                 []Line `json:"lines"`
}

func (r PlanRequest) Valid() bool {
	if !ValidID(r.PurchaseOrderID) || !ValidID(r.CommandID) || !ValidID(r.FactoryOrganizationID) || !ValidText(r.DemandReference, 1024) || r.PolicyCode != PolicyCode || !ValidSHA(r.EvidenceSHA256) || len(r.Lines) < 1 || len(r.Lines) > 32 {
		return false
	}
	ids := map[string]bool{}
	variants := map[string]bool{}
	total := 0
	for _, l := range r.Lines {
		if !ValidID(l.ID) || !ValidID(l.VariantID) || l.Quantity < 1 || l.Quantity > 1000 || ids[l.ID] || variants[l.VariantID] {
			return false
		}
		ids[l.ID] = true
		variants[l.VariantID] = true
		total += l.Quantity
	}
	return total <= 1000
}

type Command struct {
	PurchaseOrderID     string   `json:"purchase_order_id"`
	CommandID           string   `json:"command_id"`
	ExpectedVersion     int64    `json:"expected_version,string"`
	Kind                string   `json:"kind"`
	EvidenceSHA256      string   `json:"evidence_sha256"`
	LineID              string   `json:"line_id,omitempty"`
	UnitID              string   `json:"unit_id,omitempty"`
	SerialNumber        string   `json:"serial_number,omitempty"`
	VIN                 string   `json:"vin,omitempty"`
	BatterySerialNumber string   `json:"battery_serial_number,omitempty"`
	TargetState         string   `json:"target_state,omitempty"`
	ShipmentID          string   `json:"shipment_id,omitempty"`
	Units               []string `json:"units,omitempty"`
	Reason              string   `json:"reason,omitempty"`
}

func (r Command) Valid() bool {
	if !ValidID(r.PurchaseOrderID) || !ValidID(r.CommandID) || r.ExpectedVersion < 1 || !ValidSHA(r.EvidenceSHA256) {
		return false
	}
	allowed := map[string]bool{}
	switch r.Kind {
	case "submit", "confirm", "start", "cancel":
	case "register":
		if !ValidID(r.LineID) || !ValidText(r.SerialNumber, 128) || r.VIN != "" && !ValidText(r.VIN, 128) || r.BatterySerialNumber != "" && !ValidText(r.BatterySerialNumber, 128) {
			return false
		}
		allowed["line_id"] = true
		allowed["serial_number"] = true
		allowed["vin"] = true
		allowed["battery_serial_number"] = true
	case "milestone":
		if !ValidID(r.UnitID) || !map[string]bool{"assembly": true, "quality": true, "released": true, "rejected": true}[r.TargetState] {
			return false
		}
		allowed["unit_id"] = true
		allowed["target_state"] = true
		if r.TargetState == "released" || r.TargetState == "rejected" {
			allowed["reason"] = true
			if !ValidText(r.Reason, 2000) {
				return false
			}
		}
	case "ship", "receive":
		if !ValidID(r.ShipmentID) || len(r.Units) < 1 || len(r.Units) > 100 {
			return false
		}
		seen := map[string]bool{}
		for _, u := range r.Units {
			if !ValidID(u) || seen[u] {
				return false
			}
			seen[u] = true
		}
		allowed["shipment_id"] = true
		allowed["units"] = true
	case "quality", "quality-reject", "reinspect":
		if !ValidID(r.UnitID) {
			return false
		}
		allowed["unit_id"] = true
		allowed["reason"] = true
		if !ValidText(r.Reason, 2000) {
			return false
		}
	default:
		return false
	}
	values := map[string]string{"line_id": r.LineID, "unit_id": r.UnitID, "serial_number": r.SerialNumber, "vin": r.VIN, "battery_serial_number": r.BatterySerialNumber, "target_state": r.TargetState, "shipment_id": r.ShipmentID, "reason": r.Reason}
	for k, v := range values {
		if v != "" && !allowed[k] {
			return false
		}
	}
	return len(r.Units) == 0 || allowed["units"]
}

type Receipt struct {
	PurchaseOrderID string          `json:"purchase_order_id"`
	CommandID       string          `json:"command_id"`
	Version         int64           `json:"version,string"`
	Kind            string          `json:"kind"`
	Actor           string          `json:"actor"`
	RequestSHA256   string          `json:"request_sha256"`
	PayloadSHA256   string          `json:"payload_sha256"`
	Payload         json.RawMessage `json:"payload"`
	RecordedAt      time.Time       `json:"recorded_at"`
	Replay          bool            `json:"replay"`
}
type Plan struct {
	PurchaseOrderID           string   `json:"purchase_order_id"`
	DestinationOrganizationID string   `json:"destination_organization_id"`
	FactoryOrganizationID     string   `json:"factory_organization_id"`
	SupplierID                string   `json:"supplier_id"`
	State                     string   `json:"state"`
	PurchaseVersion           int64    `json:"purchase_version,string"`
	Currency                  string   `json:"currency"`
	TotalMinorUnits           int64    `json:"total_minor_units,string"`
	Version                   int64    `json:"version,string"`
	PolicyCode                string   `json:"policy_code"`
	DemandReference           string   `json:"demand_reference"`
	Lines                     []Line   `json:"lines"`
	Units                     []Unit   `json:"units"`
	Latest                    *Receipt `json:"latest,omitempty"`
	NextUnitID                string   `json:"next_unit_id,omitempty"`
}
type Unit struct {
	FactoryReviewState string `json:"factory_review_state,omitempty"`
	FactoryRequester   string `json:"factory_requester,omitempty"`
	ReceiptReviewState string `json:"receipt_review_state,omitempty"`
	ReceiptRequester   string `json:"receipt_requester,omitempty"`
	ID                 string `json:"id"`
	LineID             string `json:"line_id"`
	State              string `json:"state"`
	SerialNumber       string `json:"serial_number"`
	StockUnitID        string `json:"stock_unit_id,omitempty"`
	StockState         string `json:"stock_state,omitempty"`
	ShipmentID         string `json:"shipment_id,omitempty"`
}

File: src/platform/supply/contract.ts
// AUTHORED transport/form binding to the existing Go serial supply owner.
import {z} from "zod";
export const supplyID=z.string().regex(/^[A-Za-z0-9][A-Za-z0-9._:-]{0,127}$/u);
export const supplySHA=z.string().regex(/^[0-9a-f]{64}$/u);
const unsigned=z.string().regex(/^(0|[1-9][0-9]{0,18})$/u).refine(v=>BigInt(v)<=9223372036854775807n);
const positive=unsigned.refine(v=>v!=="0");
const text=(max:number)=>z.string().min(1).refine(v=>new TextEncoder().encode(v).length<=max&&v.trim()===v&&!/[\p{Cc}]/u.test(v)&&new TextDecoder().decode(new TextEncoder().encode(v))===v);
export const supplyPolicy="strict-serial-reference/v1";
export const supplyKinds=["submit","confirm","start","cancel","register","milestone","ship","receive","quality","quality-reject","reinspect"] as const;
const line=z.object({id:supplyID,variant_id:supplyID,quantity:z.number().int().min(1).max(1000)}).strict();
const planFields={factory_organization_id:supplyID,demand_reference:text(1024),policy_code:z.literal(supplyPolicy),lines:z.array(line).min(1).max(32).refine(v=>new Set(v.map(l=>l.id)).size===v.length&&new Set(v.map(l=>l.variant_id)).size===v.length&&v.reduce((n,l)=>n+l.quantity,0)<=1000)};
const common={organization_id:supplyID,purchase_order_id:supplyID,command_id:supplyID,evidence_sha256:supplySHA};
const base={...common,expected_version:positive};
const simple=(kind:"submit"|"confirm"|"start"|"cancel")=>z.object({kind:z.literal(kind),...base}).strict();
export const supplyCommand=z.discriminatedUnion("kind",[
 z.object({kind:z.literal("create"),...common,...planFields,destination_organization_id:supplyID,supplier_id:supplyID,currency:z.string().regex(/^[A-Z]{3}$/u),total_minor_units:unsigned}).strict().refine(v=>v.organization_id===v.destination_organization_id),
 z.object({kind:z.literal("plan"),...common,...planFields}).strict(),
 simple("submit"),simple("confirm"),simple("start"),simple("cancel"),
 z.object({kind:z.literal("register"),...base,line_id:supplyID,serial_number:text(128),vin:text(128).optional(),battery_serial_number:text(128).optional()}).strict(),
 z.object({kind:z.literal("milestone"),...base,unit_id:supplyID,target_state:z.enum(["assembly","quality","released","rejected"]),reason:text(2000).optional()}).strict().refine(v=>["released","rejected"].includes(v.target_state)?v.reason!==undefined:v.reason===undefined),
 ...(["ship","receive"] as const).map(kind=>z.object({kind:z.literal(kind),...base,shipment_id:supplyID,units:z.array(supplyID).min(1).max(100).refine(v=>new Set(v).size===v.length)}).strict()),
 ...(["quality","quality-reject","reinspect"] as const).map(kind=>z.object({kind:z.literal(kind),...base,unit_id:supplyID,reason:text(2000)}).strict())
]);
export type SupplyCommand=z.infer<typeof supplyCommand>;
export const supplyReceipt=z.object({purchase_order_id:supplyID,command_id:supplyID,version:positive,kind:z.enum(["planned",...supplyKinds]),actor:z.string().min(1).max(256),request_sha256:supplySHA,payload_sha256:supplySHA,payload:z.unknown(),recorded_at:z.iso.datetime({offset:true}),replay:z.boolean()}).strict();
export type SupplyReceipt=z.infer<typeof supplyReceipt>;
const unit=z.object({factory_review_state:z.enum(["pending","approved","rejected"]).optional(),factory_requester:z.string().min(1).max(256).optional(),receipt_review_state:z.enum(["pending","approved","rejected"]).optional(),receipt_requester:z.string().min(1).max(256).optional(),id:supplyID,line_id:supplyID,state:z.enum(["planned","assembly","quality","released","rejected","shipped","received"]),serial_number:text(128),stock_unit_id:supplyID.optional(),stock_state:z.enum(["in-transit","quarantine","available","reserved","sold","service","retired"]).optional(),shipment_id:supplyID.optional()}).strict();
export const supplyPlan=z.object({purchase_order_id:supplyID,destination_organization_id:supplyID,factory_organization_id:supplyID,supplier_id:supplyID,state:z.enum(["draft","submitted","accepted","in-production","shipped","received","cancelled"]),purchase_version:positive,version:positive,currency:z.string().regex(/^[A-Z]{3}$/u),total_minor_units:unsigned,policy_code:z.literal(supplyPolicy),demand_reference:text(1024),lines:z.array(line).min(1).max(32),units:z.array(unit).max(100),latest:supplyReceipt.optional(),next_unit_id:supplyID.optional()}).strict();
export type SupplyPlan=z.infer<typeof supplyPlan>;
export const pendingSupply=z.object({purchase_order_id:supplyID,command_id:supplyID,kind:supplyReceipt.shape.kind,version:positive,request_sha256:supplySHA}).strict();
export type PendingSupply=z.infer<typeof pendingSupply>;
export function supplyPermission(kind:SupplyCommand["kind"]){return ["confirm","start","register","milestone","ship"].includes(kind)?"supply:factory":kind==="receive"?"supply:receive":["quality","quality-reject"].includes(kind)?"supply:release":kind==="reinspect"?"supply:inspect":"supply:plan"}
export function supplySurface(kind:SupplyCommand["kind"]){return supplyPermission(kind)==="supply:factory"?"factory":"franchise"}
export function supplyPayload(c:SupplyCommand){const{organization_id:_,...body}=c;if(body.kind==="create"||body.kind==="plan"){const{kind:__,...plan}=body;return plan}return body}
// Narrow Go encoding/json correspondence: ASCII keys, validated strings and
// whole quantities1..1000 only. Not a general-purpose JSON canonicalizer.
export function supplyCanonical(value:unknown):string{
 const literal=(v:string)=>JSON.stringify(v).replace(/[<>&\u2028\u2029]/gu,x=>({"<":"\\u003c",">":"\\u003e","&":"\\u0026","\u2028":"\\u2028","\u2029":"\\u2029"}[x]!));
 if(value===null)return"null";if(typeof value==="string")return literal(value);
 if(typeof value==="number"&&Number.isInteger(value)&&value>=1&&value<=1000)return String(value);
 if(Array.isArray(value))return"["+value.map(supplyCanonical).join(",")+"]";
 if(typeof value==="object"){const o=value as Record<string,unknown>,keys=Object.keys(o).sort();if(keys.some(k=>!/^[a-z_][a-z0-9_]*$/u.test(k)))throw new Error("unsupported key");return"{"+keys.map(k=>literal(k)+":"+supplyCanonical(o[k])).join(",")+"}"}
 throw new Error("unsupported supply input");
}
export async function supplyDigest(value:string|Uint8Array){const bytes=typeof value==="string"?new TextEncoder().encode(value):value;return Array.from(new Uint8Array(await crypto.subtle.digest("SHA-256",new Uint8Array(bytes).buffer)),v=>v.toString(16).padStart(2,"0")).join("")}
export async function supplyReference(c:SupplyCommand):Promise<PendingSupply>{return pendingSupply.parse({purchase_order_id:c.purchase_order_id,command_id:c.command_id,kind:c.kind==="create"||c.kind==="plan"?"planned":c.kind,version:"expected_version"in c?(BigInt(c.expected_version)+1n).toString():"1",request_sha256:await supplyDigest(supplyCanonical(supplyPayload(c)))})}
export function supplyMatches(v:SupplyReceipt,p:PendingSupply,subject:string){return v.purchase_order_id===p.purchase_order_id&&v.command_id===p.command_id&&v.actor===subject&&v.kind===p.kind&&v.version===p.version&&v.request_sha256===p.request_sha256}
